plugin/core/input_handlers.py

```diff
@@ -0,0 +1,205 @@
+
+    def __init__(
+        self, window: sublime.Window, initial_value: Optional[Union[str, sublime.ListInputItem]] = None
+    ) -> None:
+        super().__init__()
+        self._window = window
+        self._initial_value = initial_value
+
+    @final
+    def list_items(self) -> ListItemsReturn:
+        if self._initial_value is not None:
+            sublime.set_timeout(self._select_and_reset)
+            return [self._initial_value], 0  # pyright: ignore[reportGeneralTypeIssues]
+        else:
+            return self.get_list_items()
+
+    def _select_and_reset(self) -> None:
+        self._initial_value = None
+        if self._window.is_valid():
+            self._window.run_command('select')
+
+    @abstractmethod
+    def get_list_items(self) -> ListItemsReturn:
+        raise NotImplementedError()
+
+
+class DynamicListInputHandler(sublime_plugin.ListInputHandler, metaclass=ABCMeta):
+    """ A ListInputHandler which can update its items while typing in the input field.
+
+    Subclasses of PreselectedListInputHandler must not implement the `list_items` method, but can override
+    `get_list_items` for the initial list items. The `on_modified` method will be called after a small delay (debounced)
+    whenever changes were made to the input text. You can use this to call the `update` method with a list of
+    `ListInputItem`s to update the list items.
+
+    To create an instance of the derived class pass the command instance and the command arguments to the constructor,
+    like this:
+
+    def input(self, args):
+        return MyDynamicListInputHandler(self, args)
+
+    For now, the type of the command must be a WindowCommand, but maybe it can be generalized later if needed.
+    This class will set and modify `_items` and '_text' attributes of the command, so make sure that those attribute
+    names are not used in another way in the command's class.
+    """
+
+    def __init__(self, command: sublime_plugin.WindowCommand, args: Dict[str, Any]) -> None:
+        super().__init__()
+        self.command = command
+        self.args = args
+        self.text = getattr(command, '_text', '')
+        self.listener = None  # type: Optional[sublime_plugin.TextChangeListener]
+        self.input_view = None  # type: Optional[sublime.View]
+
+    def attach_listener(self) -> None:
+        for buffer in sublime._buffers():  # type: ignore
+            view = buffer.primary_view()
+            # This condition to find the input field view might not be sufficient if there is another command palette
+            # open in another group in the same window
+            if view.element() == 'command_palette:input' and view.window() == self.command.window:
+                self.input_view = view
+                break
+        else:
+            raise RuntimeError('Could not find the Command Palette input field view')
+        self.listener = InputListener(self)
+        self.listener.attach(buffer)
+        if ST_VERSION < 4161:
```

Comment:
I would not commit this workaround if it was fixed in a later version of ST.

The benefit is that we avoid having slightly different behavior that this if introduces.
if someone opens an issue, we can just tell them that it was fixed in a later version of ST.

Wdyt?

Reply:
Hm, are you sure? The problem is that the feature is basically unusable without this workaround on old ST builds. It would select the whole input text when it gets refreshed, so if you continue typing everything would be deleted again. 

I would be fine to remove it if we modify the tag prefix for new releases and add a new version restriction for Package Control at https://github.com/sublimelsp/repository/blob/200d5081fdbbff3c2c21f85a4e84f600ec2a7a84/repository.json#L49-L52, but otherwise I would keep in in the code.

Btw, I'm quite sure that the current code uses API functions introduced after ST 4070 in a few places, so I think not all features are compatible with ST 4070 anyway. But I guess nobody cares and there are no users for such old ST4 versions anyway.

Reply:
> Hm, are you sure?

I am ok with merging the code as is.
I just wanted to check if you think that it would be good to remove the code in the `ST_VERSION < 4161:` branches,
to keep the code simpler.

(I think that in one or two years, 
the code if `ST_VERSION < 4161:` will become obsolete and not necessary at some point in future.)